lib/streamlit/watcher/util.py

```diff
@@ -75,3 +69,48 @@ def _get_file_content_with_blocking_retries(file_path):
                 raise e
             time.sleep(_RETRY_WAIT_SECS)
     return content
+
+
+def _dirfiles(dir_path: str, glob_pattern: str) -> str:
+    p = Path(dir_path)
+    filenames = sorted(
+        [f.name for f in p.glob(glob_pattern) if not f.name.startswith(".")]
+    )
+    return "+".join(filenames)
+
+
+def _stable_dir_identifier(dir_path: str, glob_pattern: str = "*") -> str:
+    """Wait for the files in a directory to look stable-ish before returning an id.
+
+    We do this to deal with problems that would otherwise arise from many tools
+    (e.g. git) and editors (e.g. vim) "editing" files (from the user's
+    perspective) by doing some combination of deleting, creating, and moving
+    various files under the hood.
+
+    Because of this, we're unable to rely on FileSystemEvents that we receive
+    from watchdog to determine when a file has been added to or removed from a
+    directory.
+
+    This is a bit of an unfortunate situation, but the approach we take here is
+    most likely fine as:
+      * The worst thing that can happen taking this approach is a false
+        positive page added/removed notification, which isn't too disastrous
+        and can just be ignored.
+      * It is impossible (that is, I'm fairly certain that the problem is
+        undecidable) to know whether a file created/deleted/moved event
+        corresponds to a legitimate file creation/deletion/move or is part of
+        some sequence of events that results in what the user sees as a file
+        "edit".
+    """
+    dirfiles = _dirfiles(dir_path, glob_pattern)
+
+    for _ in range(_MAX_RETRIES):
+        time.sleep(_RETRY_WAIT_SECS)
+
+        new_dirfiles = _dirfiles(dir_path, glob_pattern)
+        if dirfiles == new_dirfiles:
+            break
+
+        dirfiles = new_dirfiles
+
+    return f"{dir_path}+{dirfiles}"
```

Comment:
This may be a premature optimization, but it feels slightly iffy to be returning this ID as a (potentially gigantic!) string which will then immediately be fed into a hash function.

- Is this a reasonable concern, or am I being paranoid about what's actually a pathological use-case that we aren't likely to see in practice?
- Is there a potential savings if `_stable_dir_identifier` does the filename hashing itself, incrementally, and returns the MD5 hash instead of a big concatenated string?

Reply:
Hm, I think it's pretty unlikely that performance due to long strings being returned from this function will be too much of an issue in practice.

There are a few reasons for this:
* Since (for now) we'll only be considering `*.py` files 1-level deep in the `pages/` directory, it's pretty unlikely that many apps will have enough pages so that this function outputs strings comparable to the length of source files that we're already hashing today (you'd probably need a few thousand pages in an app before it's comparable to an especially-long source file, which I can't see happening often)
* I'm not super familiar with hash function performance, but according to some stackoverflow answers, md5's time complexity is `O(n)` where `n` is the length of the string, so I feel like (at least asymptotically) there shouldn't be much of a difference between repeatedly hashing shorter strings and hashing one big one

Reply:
Oh yeah, to be clear I don't think this is a wasted-cpu-cycles issue, but rather a wasted-memory issue. (Since the string is only going to be fed into MD5 anyway, why bother to concat-and-then-hash, instead of just hashing incrementally in this function and save the construction of the big string.)

But I didn't realize this was strictly a 1-level deep scan, which makes me less concerned!

Reply:
👍 

I'll make a note to investigate this a bit more before the multipage apps launch (this code path should be unused until the feature branch is merged), but I think performance problems due to the length of the string being returned here should be extremely difficult to run into given the use cases in which they would appear